kroxylicious/src/test/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandlerTest.java

```diff
@@ -0,0 +1,224 @@
+                .setRequestApiVersion(apiVersion)
+                .setClientId("client-id")
+                .setCorrelationId(downstreamCorrelationId);
+
+        inboundChannel.writeInbound(new DecodedRequestFrame<>(apiVersion, corrId, true, header, body));
+    }
+
+    @BeforeEach
+    public void buildChannel() {
+        inboundChannel = new EmbeddedChannel();
+        corrId = 0;
+    }
+
+    @AfterEach
+    public void closeChannel() {
+        inboundChannel.close();
+    }
+
+    public static List<Arguments> provideArgsForExpectedFlow() {
+        var result = new ArrayList<Arguments>();
+        boolean[] tf = { true, false };
+        for (boolean sslConfigured : tf) {
+            for (boolean haProxyConfigured : tf) {
+                for (boolean saslOffloadConfigured : tf) {
+                    for (boolean sendApiVersions : tf) {
+                        for (boolean sendSasl : tf) {
+                            result.add(Arguments.of(sslConfigured, haProxyConfigured, saslOffloadConfigured, sendApiVersions, sendSasl));
+                        }
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Test the normal flow, in a number of configurations.
+     * @param sslConfigured Whether SSL is configured
+     * @param haProxyConfigured
+     * @param saslOffloadConfigured
+     * @param sendApiVersions
+     * @param sendSasl
+     */
+    @ParameterizedTest
+    @MethodSource("provideArgsForExpectedFlow")
+    public void expectedFlow(boolean sslConfigured,
+                             boolean haProxyConfigured,
+                             boolean saslOffloadConfigured,
+                             boolean sendApiVersions,
+                             boolean sendSasl) {
+
+        var dp = new SaslDecodePredicate(saslOffloadConfigured);
+        ArgumentCaptor<NetFilter.NetFilterContext> valueCapture = ArgumentCaptor.forClass(NetFilter.NetFilterContext.class);
+        var filter = mock(NetFilter.class);
+        doAnswer(i -> {
+            NetFilter.NetFilterContext ctx = i.getArgument(0);
+            assertEquals(sslConfigured ? SNI_HOSTNAME : null, ctx.sniHostname());
+            if (haProxyConfigured) {
+                assertEquals("embedded", String.valueOf(ctx.srcAddress()));
+                assertEquals("1.2.3.4", ctx.clientAddress());
+            }
+            else {
+                assertEquals("embedded", String.valueOf(ctx.srcAddress()));
+                assertEquals("embedded", ctx.clientAddress());
+            }
+            assertEquals(sendApiVersions ? "foo" : null, ctx.clientSoftwareName());
+            assertEquals(sendApiVersions ? "1.0.0" : null, ctx.clientSoftwareVersion());
+            if (saslOffloadConfigured && sendSasl) {
+                assertEquals("alice", ctx.authorizedId());
+            }
+            else {
+                assertNull(ctx.authorizedId());
+            }
```

Comment:
Shouldn't these be tests of the context directly? Possibly hinting that context should be a first class thing in its own right. 

Given a context with a channel in X state assert that the expected values are returned. Asserting that as a side effect of this test feels clunky and overly complicated.